Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/ComponentModeSwitcher.cpp

```diff
@@ -0,0 +1,381 @@
+                }
+            }
+        }
+    }
+
+    void ComponentModeSwitcher::RemoveExclusiveComponents(const AZ::Entity& entity)
+    {
+        // Get all component Ids associated with entity, identify which ones to remove and then remove them
+        AZStd::set<AZ::TypeId> componentIds;
+        for (const auto& entityComponent : entity.GetComponents())
+        {
+            componentIds.insert(entityComponent->RTTI_GetType());
+        }
+
+        for (AZStd::vector<ComponentData>::reverse_iterator it = m_addedComponents.rbegin(); it != m_addedComponents.rend(); ++it)
+        {
+            if(!componentIds.contains(it->m_component->RTTI_GetType()))
+            {
+                RemoveComponentButton(it->m_pairId);
+            }
+        }
+    }
+
+    void ComponentModeSwitcher::OnEntityComponentDisabled(const AZ::EntityId& entityId, const AZ::ComponentId& componentId)
+    {
+        auto pairId = AZ::EntityComponentIdPair(entityId, componentId);
+        RemoveComponentButton(pairId);
+    }
+
+    void ComponentModeSwitcher::OnEntityComponentEnabled(const AZ::EntityId& entityId, const AZ::ComponentId& componentId)
+    {
+        auto pairId = AZ::EntityComponentIdPair(entityId, componentId);
+        AddComponentButton(pairId);
+    }
+
+    void ComponentModeSwitcher::OnEntityComponentAdded(
+        [[maybe_unused]] const AZ::EntityId& entity, [[maybe_unused]] const AZ::ComponentId& component)
+    {
+        // when a component is added to an entity set a staging member variable to the entityComponentIdPair
+        m_componentModePair = AZ::EntityComponentIdPair(entity, component);
+        m_addOrRemove = AddOrRemoveComponent::Add;
+    }
+
+    void ComponentModeSwitcher::OnEntityComponentRemoved(
+        [[maybe_unused]] const AZ::EntityId& entityId, [[maybe_unused]] const AZ::ComponentId& componentId)
+    {
+        m_componentModePair = AZ::EntityComponentIdPair(entityId, componentId);
+        m_addOrRemove = AddOrRemoveComponent::Remove;
+    }
+
+    // this is called twice when a component is added, once while the component is pending and
+    // once when it has been added fully to the entity. This is handled in UpdateComponentButton
+    void ComponentModeSwitcher::OnEntityCompositionChanged(const AzToolsFramework::EntityIdList& entityIdList)
+    {
+        if (AZStd::find(entityIdList.begin(), entityIdList.end(), m_componentModePair.GetEntityId()) != entityIdList.end())
+        {
+            if (m_addOrRemove == AddOrRemoveComponent::Add)
+            {
+                AddComponentButton(m_componentModePair);
+            }
+            else
+            {
+                RemoveComponentButton(m_componentModePair);
+            }
+        }
+    }
+
+    void ComponentModeSwitcher::AfterEntitySelectionChanged(
+        [[maybe_unused]] const EntityIdList& newlySelectedEntities, [[maybe_unused]] const EntityIdList& newlyDeselectedEntities)
```

Comment:
update: Please remove `[[maybe_unused]]` as the parameters are referenced in `UpdateSwitcherOnEntitySelectionChange`.